src/include/daos_srv/ad_mem.h

```diff
@@ -0,0 +1,69 @@
+#ifndef __DAOS_AD_HOC_MEM_H__
+#define __DAOS_AD_HOC_MEM_H__
+
+#include <daos/mem.h>
+
+/* the memory region (blob) manged by ad-hoc allocator */
+struct ad_blob;
+
+/* Ad-hoc mem transaction handle */
+struct ad_tx {
+	struct ad_blob		*tx_blob;
+	uint64_t		 tx_id;
+	bool			 tx_redo_only;
+	d_list_t		 tx_redo;
+	d_list_t		 tx_undo;
+};
+
+/** start a ad-hoc memory transaction */
+int
+ad_tx_begin(struct ad_blob *blob, struct ad_tx *tx);
+
+/** complete a ad-hoc memory transaction */
+int
+ad_tx_end(struct ad_tx *tx, int err);
+
+/**
+ * declare this storage region is going to be changed (for redo and undo), it does memset
+ * the region if @reset is true.
+ */
+int
+ad_tx_touch_region(void *addr, daos_size_t size, bool reset, struct ad_tx *tx);
```

Comment:
minor, may keep the tx as first arg same as other APIs.

Reply:
indeed.